tests/tests/classes/PluginManager.js

```diff
@@ -0,0 +1,299 @@
+
+      // used to test if the function was executed correctly
+      this._deployedFunctions = 0;
+      this._deployedResources = 0;
+    }
+
+    functions() {
+      return new Promise((resolve, reject) => {
+        this._deployedFunctions = this._deployedFunctions + 1;
+        return resolve();
+      });
+    }
+
+    resources() {
+      return new Promise((resolve, reject) => {
+        this._deployedResources = this._deployedResources + 1;
+        return resolve();
+      });
+    }
+  }
+
+  class SynchronousPluginMock {
+    constructor() {
+      this.commands = {
+        deploy: {
+          usage: 'Deploy to the default infrastructure',
+          lifeCycleEvents: [
+            'resources',
+            'functions'
+          ],
+          commands: {
+            onpremises: {
+              usage: 'Deploy to your On-Premises infrastructure',
+              lifeCycleEvents: [
+                'resources',
+                'functions'
+              ],
+            },
+          },
+        },
+      };
+
+      this.hooks = {
+        'deploy:functions': this.functions.bind(this),
+        'deploy:onpremises:beforeFunctions': this.resources.bind(this),
+      };
+
+      // used to test if the function was executed correctly
+      this._deployedFunctions = 0;
+      this._deployedResources = 0;
+    }
+
+    functions() {
+      this._deployedFunctions = this._deployedFunctions + 1;
+    }
+
+    resources() {
+      this._deployedResources = this._deployedResources + 1;
+    }
+  }
+
+  beforeEach(() => {
+    serverless = new Serverless({});
+    pluginManager = new PluginManager(serverless);
+  });
+
+  describe('#constructor()', () => {
+    it('should set the serverless instance', () => {
+      expect(pluginManager._serverless).to.deep.equal(serverless);
+    });
+
+    it('should create an empty pluginInstances array', () => {
+      expect(pluginManager._pluginInstances.length).to.equal(0);
+    });
+
+    it('should create an empty commandsList array', () => {
+      expect(pluginManager._commandsList.length).to.equal(0);
+    });
+
+    it('should create an empty commands object', () => {
+      expect(pluginManager._commands).to.deep.equal({});
+    });
+  });
+
+  describe('#addPlugin()', () => {
+    it('should add a plugin instance to the pluginInstances array', () => {
+      pluginManager._addPlugin(SynchronousPluginMock);
+
+      expect(pluginManager._pluginInstances[0]).to.be.an.instanceof(SynchronousPluginMock);
+    });
+
+    it('should load the plugin commands', () => {
+      pluginManager._addPlugin(SynchronousPluginMock);
+
+      expect(pluginManager._commandsList[0]).to.have.property('deploy');
+    });
+  });
+
+  describe('#loadAllPlugins()', () => {
+    describe('when loading all plugins', () => {
+      it('should load all plugins when no service plugins are given', () => {
+        pluginManager.loadAllPlugins();
+
+        expect(pluginManager._pluginInstances.length).to.be.at.least(1);
```

Comment:
Can you check for the exact plugin that is stored in here, because this doesn't test that its actually only loading the core plugin, but just tests that any plugin is loaded. E.g. if the implementation loads something else accidentally this could still be valid